### PR TITLE
Heifer ration

### LINKED ISSUE #69: Pen Allocation Error
If the number of stalls in the pens for lactating cows is fewer than the herd number, I get an error and the model crashes.

### LINKED ISSUE #74: Pen change Frequency
Cows change pens very frequently. Is this what we want? What is causing them to change so frequently?

### PR DESCRIPTION
**Pull request by cvankerkhove**
[Pen Allocation Flow Chart.pdf](https://github.com/RuminantFarmSystems/MASM/files/6480760/Pen.Allocation.Flow.Chart.pdf)
This pull request includes major development from 2 branches (heifer_raiton and v_patches). The Heifer ration development mainly concerns updates to the ration module and optimization process, along with an overhaul of the requirements structure. The v_patches development mainly contributes an updated pen allocation method. A document is attached with a more in depth analysis of the changes made in this pull request. 
[Heifer Ration Pull Request.docx](https://github.com/RuminantFarmSystems/MASM/files/6461773/Heifer.Ration.Pull.Request.docx)

### CODE REVIEW

**Comment by militsasotirova** [comment]
Could you add pyomo (and remove PuLP as we don't use that package anymore) to the requirements.txt file?

**Comment by KFosterReed** [comment]
So I am trying to run a simulation with a different set of feeds and am getting a key error for a feed that is not supposed to be in the available feeds but is listed in the hardcoded_ration.py ration. It looks like the calf_ration.py is still calling the get_ration() function from the hardcoded_ration.py. Also, the calf_ration.py file doesn't have an introduction or author listed... do we know who wrote this code? @militsasotirova , we will definitely need to be using the calf_ration.calc_requirements() function to get the P_balance working as well. A side note is that we should probably split this function up into one that calculates requirements and one that calculates intake for calves. Here is the traceback for the error: 

Traceback_5-17

I will share the feed json file in Slack. 

I'm trying to think about how to move forward with this pull request since the calf ration isn't technically part of it. The calf ration should be using it's own set of feeds since we definitely don't want milk replacer or calf starter feed to be available to all animals. Could we add a separate input in the feeds .json that is 'available_feeds_calf' and have those listed in the hardcoded_ration.py? My thought is that if we go that route, we won't have to review the calf nutrient requirements function. @cvankerkhove, what do you think about how having a separate available_feeds object would work in the ration driver structure?

**Comment by militsasotirova** [comment]
> So I am trying to run a simulation with a different set of feeds and am getting a key error for a feed that is not supposed to be in the available feeds but is listed in the hardcoded_ration.py ration. It looks like the calf_ration.py is still calling the get_ration() function from the hardcoded_ration.py. Also, the calf_ration.py file doesn't have an introduction or author listed... do we know who wrote this code? @militsasotirova , we will definitely need to be using the calf_ration.calc_requirements() function to get the P_balance working as well. A side note is that we should probably split this function up into one that calculates requirements and one that calculates intake for calves. Here is the traceback for the error:
> 
> Traceback_5-17
> 
> I will share the feed json file in Slack.
> 
> I'm trying to think about how to move forward with this pull request since the calf ration isn't technically part of it. The calf ration should be using it's own set of feeds since we definitely don't want milk replacer or calf starter feed to be available to all animals. Could we add a separate input in the feeds .json that is 'available_feeds_calf' and have those listed in the hardcoded_ration.py? My thought is that if we go that route, we won't have to review the calf nutrient requirements function. @cvankerkhove, what do you think about how having a separate available_feeds object would work in the ration driver structure?

I think it's a good idea to separate the logic into two functions, and we should probably call the method calc_ration() rather than optimize() since we're not doing any optimization to get that ration. @katrina433 wrote the code in the calf ration file and connected it to the rest of the module. Another idea for when to make changes is since the calf ration isn't part of the functionality of this branch, maybe we can put these changes in the next life cycle clean up branch (we already have the functionality - its just a matter of reorganization) that I think you mentioned Manfei is making? In order to test this branch without running into that error, you have to modify the hardcoded ration dictionary to have the same keys as the feeds you put in the JSON file.

**Comment by KFosterReed** [comment]
> > So I am trying to run a simulation with a different set of feeds and am getting a key error for a feed that is not supposed to be in the available feeds but is listed in the hardcoded_ration.py ration. It looks like the calf_ration.py is still calling the get_ration() function from the hardcoded_ration.py. Also, the calf_ration.py file doesn't have an introduction or author listed... do we know who wrote this code? @militsasotirova , we will definitely need to be using the calf_ration.calc_requirements() function to get the P_balance working as well. A side note is that we should probably split this function up into one that calculates requirements and one that calculates intake for calves. Here is the traceback for the error:
> > Traceback_5-17
> > I will share the feed json file in Slack.
> > I'm trying to think about how to move forward with this pull request since the calf ration isn't technically part of it. The calf ration should be using it's own set of feeds since we definitely don't want milk replacer or calf starter feed to be available to all animals. Could we add a separate input in the feeds .json that is 'available_feeds_calf' and have those listed in the hardcoded_ration.py? My thought is that if we go that route, we won't have to review the calf nutrient requirements function. @cvankerkhove, what do you think about how having a separate available_feeds object would work in the ration driver structure?
> 
> I think it's a good idea to separate the logic into two functions, and we should probably call the method calc_ration() rather than optimize() since we're not doing any optimization to get that ration. @katrina433 wrote the code in the calf ration file and connected it to the rest of the module. Another idea for when to make changes is since the calf ration isn't part of the functionality of this branch, maybe we can put these changes in the next life cycle clean up branch (we already have the functionality - its just a matter of reorganization) that I think you mentioned Manfei is making? In order to test this branch without running into that error, you have to modify the hardcoded ration dictionary to have the same keys as the feeds you put in the JSON file.





Great ideas, Militsa! I changed the hardcoded ration and the heifer ration is now working. I think having the calf ration updates in Manfei's next life-cycle cleanup branch sounds like a good plan. @katrina433 , does that work for you? Will you be available early on in the summer to do this?

**Comment by katrina433** [comment]
> > > So I am trying to run a simulation with a different set of feeds and am getting a key error for a feed that is not supposed to be in the available feeds but is listed in the hardcoded_ration.py ration. It looks like the calf_ration.py is still calling the get_ration() function from the hardcoded_ration.py. Also, the calf_ration.py file doesn't have an introduction or author listed... do we know who wrote this code? @militsasotirova , we will definitely need to be using the calf_ration.calc_requirements() function to get the P_balance working as well. A side note is that we should probably split this function up into one that calculates requirements and one that calculates intake for calves. Here is the traceback for the error:
> > > Traceback_5-17
> > > I will share the feed json file in Slack.
> > > I'm trying to think about how to move forward with this pull request since the calf ration isn't technically part of it. The calf ration should be using it's own set of feeds since we definitely don't want milk replacer or calf starter feed to be available to all animals. Could we add a separate input in the feeds .json that is 'available_feeds_calf' and have those listed in the hardcoded_ration.py? My thought is that if we go that route, we won't have to review the calf nutrient requirements function. @cvankerkhove, what do you think about how having a separate available_feeds object would work in the ration driver structure?
> > 
> > 
> > I think it's a good idea to separate the logic into two functions, and we should probably call the method calc_ration() rather than optimize() since we're not doing any optimization to get that ration. @katrina433 wrote the code in the calf ration file and connected it to the rest of the module. Another idea for when to make changes is since the calf ration isn't part of the functionality of this branch, maybe we can put these changes in the next life cycle clean up branch (we already have the functionality - its just a matter of reorganization) that I think you mentioned Manfei is making? In order to test this branch without running into that error, you have to modify the hardcoded ration dictionary to have the same keys as the feeds you put in the JSON file.
> 
> Great ideas, Militsa! I changed the hardcoded ration and the heifer ration is now working. I think having the calf ration updates in Manfei's next life-cycle cleanup branch sounds like a good plan. @katrina433 , does that work for you? Will you be available early on in the summer to do this?

Yes I can work on that early this summer.

**Comment by cvankerkhove** [comment]
> > So I am trying to run a simulation with a different set of feeds and am getting a key error for a feed that is not supposed to be in the available feeds but is listed in the hardcoded_ration.py ration. It looks like the calf_ration.py is still calling the get_ration() function from the hardcoded_ration.py. Also, the calf_ration.py file doesn't have an introduction or author listed... do we know who wrote this code? @militsasotirova , we will definitely need to be using the calf_ration.calc_requirements() function to get the P_balance working as well. A side note is that we should probably split this function up into one that calculates requirements and one that calculates intake for calves. Here is the traceback for the error:
> > Traceback_5-17
> > I will share the feed json file in Slack.
> > I'm trying to think about how to move forward with this pull request since the calf ration isn't technically part of it. The calf ration should be using it's own set of feeds since we definitely don't want milk replacer or calf starter feed to be available to all animals. Could we add a separate input in the feeds .json that is 'available_feeds_calf' and have those listed in the hardcoded_ration.py? My thought is that if we go that route, we won't have to review the calf nutrient requirements function. @cvankerkhove, what do you think about how having a separate available_feeds object would work in the ration driver structure?
> 
> I think it's a good idea to separate the logic into two functions, and we should probably call the method calc_ration() rather than optimize() since we're not doing any optimization to get that ration. @katrina433 wrote the code in the calf ration file and connected it to the rest of the module. Another idea for when to make changes is since the calf ration isn't part of the functionality of this branch, maybe we can put these changes in the next life cycle clean up branch (we already have the functionality - its just a matter of reorganization) that I think you mentioned Manfei is making? In order to test this branch without running into that error, you have to modify the hardcoded ration dictionary to have the same keys as the feeds you put in the JSON file.

Yes i think that all sounds great, in reality we probably should have calf_optimize separate as the ration_driver is not built at all to handle calves. But yeah, until then we just need to change the keys in the hardcoded ration. Hopefully with the help of Katrina we can finally remove the hard_coded.ration.py for good with changes!

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 1]
```diff
@@ -1,27 +1,33 @@
...
 
 import math
 
 
-def calc_rqmts(BW, MW, DOP, parity, CI, TP_Milk, Fat_Milk, Lactose_Milk,
-               Milk, DIM, lactating):
+def calc_rqmts(BW, MW, DOP, animal_type, parity=None, CI=None , TP_Milk=None, Fat_Milk=None,
+                Lactose_Milk=None, Milk=None, DIM=None, lactating=None,
+                BCS5= None, PrevTemp = None,
+                ADG_heifer = None, Age = None):
```
Comment (militsasotirova) [inline RUFAS/routines/animal/ration/animal_requirements.py:15]: Pycharm shows that the Age parameter is unused in the method, so it could be removed

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 2]
```diff
@@ -39,179 +50,232 @@ def calc_rqmts(BW, MW, DOP, parity, CI, TP_Milk, Fat_Milk, Lactose_Milk,
...
+        # [A.Cow.C.1]
+        # Calcium maintenance requirement (g)
+        if lactating:
+            Ca_maint = 0.031 * BW + 0.08 * (BW / 100)
+        else:
+            Ca_maint = 0.0154 * BW + 0.08 * (BW / 100)
+    elif animal_type == 'heifer':
+        # [A.Heifer.C.1]
+        # Calcium maintenance requirement (g)
+        Ca_main = 0.0154*BW + 0.08*(BW/100)
```
Comment (militsasotirova) [inline RUFAS/routines/animal/ration/animal_requirements.py:208]: Should line 208 be Ca_maint? Ca_main is used on line 232 and I'm not sure if they're two different variables - if they are maybe we should add '_heifer' instead

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 3]
```diff
@@ -231,13 +295,13 @@ def energy_activity_rqmts(BW, housing, distance):
     """
     # Activity requirements
     # ---------------------
-    # [A.Cow.A.4]
+    # [A.Cow.A.4]-[A.Heifer.A.5]
     # Net energy for activity requirement caused by grazing system (Mcal)
     if housing == 'Grazing':
         NEa1 = 0.0012 * BW
```
Comment (militsasotirova) [inline RUFAS/routines/animal/ration/animal_requirements.py:301]: A more descriptive name for NEa1 would probably be something like NEa_grazing

File: RUFAS/routines/animal/animal_management.py [hunk 4]
```diff
@@ -333,8 +347,20 @@ def calc_nutrient_rqmts(self, feed, temp):
             feed: instance of the feed class
             temp: the temperature on the current day
         """
-        for pen in self.all_pens:
-            pen.call_animal_nutrient_rqmts(feed, temp)
+        for calf in self.calves:
+            calf.calc_nutrient_rqmts(feed, temp)
```
Comment (militsasotirova) [inline RUFAS/routines/animal/animal_management.py:351]: When Katrina makes the changes discussed earlier, we should probably rename the calf method to set_nutrient_rqmts for consistency with the other animal classes

File: RUFAS/routines/animal/animal_management.py [hunk 5]
```diff
@@ -647,6 +818,7 @@ def daily_updates(self, feed, weather, time):
             weather: instance of the Weather class defined in classes.py
             time: instance of the Time class defined in classes.py
         """
+        print(self.simulation_day)
```
Comment (militsasotirova) [inline RUFAS/routines/animal/animal_management.py:821]: We should remove this line before merging

File: RUFAS/routines/animal/animal_management.py [hunk 6]
```diff
@@ -365,128 +391,284 @@ def daily_update_id_pen(self, animals_added, ids_removed, calves_born, feed, tem
...
+        #density per-pen for even distribution
+        density = (len(self.heiferIIIs) + len(dry_cows)) / sum(stalls)
+        group = []
+        #grouping by heiferIIs first
+        for hef3 in self.heiferIIIs:
+            if len(group)/close_up_pens[0].num_stalls <= density:
+                group.append(hef3)
+            else:
+                close_up_pens[0].update_animals(group, 'close-up')
+                close_up__pens.pop(0)
```
Comment (militsasotirova) [inline RUFAS/routines/animal/animal_management.py:646]: There are two underscores on line 646

File: RUFAS/routines/animal/animal_management.py [hunk 7]
```diff
@@ -365,128 +391,284 @@ def daily_update_id_pen(self, animals_added, ids_removed, calves_born, feed, tem
...
-            self.all_pens[1].update_animals(self.heiferIs)
-            self.all_pens[2].update_animals(heifers)
-        else:
-            self.all_pens[0].update_animals(self.calves)
-            self.all_pens[1].update_animals(self.heiferIs)
-            self.all_pens[2].update_animals(self.heiferIIs)
-            self.all_pens[3].update_animals(self.heiferIIIs)
+        #TODO
+        #-mark pens after grouping
+        #-adding new animals to pens with lowest stocking density
```
Comment (militsasotirova) [inline RUFAS/routines/animal/animal_management.py:495]: It seems as though you've already done this in daily_update_id_pen(), so you could remove the TODO

**Review by militsasotirova — COMMENTED** [review]
I like the cleanup of methods and the consolidation of all the requirements calculations in one spot - its a lot cleaner now and easier to follow. In terms of the pen allocation, I like how if a user doesn’t allow any pen to have a certain animal group, the simulation creates one and the new code in daily_update_pen_id() makes a lot more sense. Overall, everything looks good (I ran the program several times with different pen configurations to try and test edge cases) and once those small changes that I commented about earlier are made, I’ll approve!

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 8]
```diff
@@ -39,179 +50,232 @@ def calc_rqmts(BW, MW, DOP, parity, CI, TP_Milk, Fat_Milk, Lactose_Milk,
...
     # Net energy for maintenance requirement (Mcal)
-    NEmaint = (0.08 * (BW - CW) ** 0.75)
+    if animal_type == 'cow':
+        NEmaint = (0.08 * (BW - CW) ** 0.75)
+
+    # Heifer Energy Requirements
+    # [A.Heifer.A.3]
+    elif animal_type == 'heifer':
+        # BCS9 = Body condition score, 1 - 9 basis
+        BCDS9 = (BCS5 -1) * 2 + 1
```
Comment (manfei-L) [inline RUFAS/routines/animal/ration/animal_requirements.py:73]: Both BCDS9 and BCS5 are only used for heifer ration and are representing body condition score, is there a particular reason we want to keep them both? Would it be more clear to just using one?

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 9]
```diff
@@ -39,179 +50,232 @@ def calc_rqmts(BW, MW, DOP, parity, CI, TP_Milk, Fat_Milk, Lactose_Milk,
...
-    else:
-        ADG = 0
-    # [A.Cow.A.12]
+    if animal_type == 'cow':
+        if parity == 1 and CI != 0:
+            ADG = ((0.92 - 0.82) * MSBW) / CI
+        elif parity == 2 and CI != 0:
+            ADG = ((1 - 0.92) * MSBW) / CI
+        else:
+            ADG = 0
```
Comment (manfei-L) [inline RUFAS/routines/animal/ration/animal_requirements.py:103]: In cow.py, we have the variable target_adg_cow, which uses the same equation. It would be better if we use that variable here from the animal instead of calculation it again. (there are some updates made in the master branch to make the ADG more precise).

File: RUFAS/routines/animal/ration/animal_requirements.py [hunk 10]
```diff
@@ -39,179 +50,232 @@ def calc_rqmts(BW, MW, DOP, parity, CI, TP_Milk, Fat_Milk, Lactose_Milk,
...
+    # [A.Heifer.A.12]
+    # Average Daily Gain (kg)
+    elif animal_type == 'heifer':
+        ADG = max(ADG_heifer, 0)
```
Comment (manfei-L) [inline RUFAS/routines/animal/ration/animal_requirements.py:107]: I think the ADG_heifer used here is the bodyweight change (including the growth and pregnancy for preg heifers); is that what is needed here?

**Review by manfei-L — COMMENTED** [review]
I like the clarity of the calculations of requirements, and the code makes a lot more sense when it can add new animals to the pen with the lowest stocking density. I followed several pen histories in the output, and the classes_in_pen works well with 5 types of pens (calf, growing, close-up, lactating, and mixed). There are some CSV files that should be removed before merging. Though when I tried the scenario with a fewer number-of-stalls for each pen in the input, I have two questions about the output: 1. there were 15 pens in the pen history, but only 13 had pen reports, the pen with the mixed animal group didn't have any pen report produced. 2. among the 13 produced pen_report, those heifer pens didn't have non-zero manure reports the same as others.
image
